### PR TITLE
Add support for capturing tensors with score_mod

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124799
* __->__ #124444

```
import torch
from torch import nn
import torch.nn.functional as F
import torch._inductor.config as config
# torch.set_default_device('cuda')

import torch
from torch.nn.attention._templated_attention import _templated_attention as templated_attention
from triton.testing import do_bench
from torch.nn.attention import SDPBackend, sdpa_kernel


index = torch.ops.aten
torch.manual_seed(0)

B = 16
H = 16
S = 2048
D = 64

head_scale = torch.randn(H, device='cuda')
def alibi(score, batch, head, token_q, token_kv):
    return score + torch.ops.aten.index(head_scale, [head]) * (token_q - token_kv)
bias = torch.randn(H, S, S, dtype=torch.float16, device='cuda')

query = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)
key = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)
value = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)

compiled = torch.compile(templated_attention)
out = compiled(query, key, value, score_mod=alibi)
out2 = templated_attention(query, key, value,score_mod=alibi)
print((out - out2).abs().mean())
assert (out - out2).abs().mean() < 1e-3
print("Flash (no mask): ", do_bench(lambda: F.scaled_dot_product_attention(query, key, value)))
print("Flash (mask): ", do_bench(lambda: F.scaled_dot_product_attention(query, key, value, attn_mask=bias)))
print("flexattention: ", do_bench(lambda: compiled(query, key, value, score_mod=alibi)))
```
image

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D56583900](https://our.internmc.facebook.com/intern/diff/D56583900)

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -1333,7 +1333,7 @@ def create_wrapped_node(
 
         def create_scalar():
             return query.call_method(
-                tx, "new_empty", (SourcelessBuilder.create(tx, []),), {}
+                tx, "new_empty", (SourcelessBuilder.create(tx, []),), {"dtype": SourcelessBuilder.create(tx, torch.int32)}
```
Comment (drisspg) [inline torch/_dynamo/variables/higher_order_ops.py:1336]: nice catch

**Review by jansel — APPROVED** [review]

File: test/inductor/test_templated_attention.py [hunk 2]
```diff
@@ -147,23 +157,116 @@ def score_mod_2(score, b, h, m, n):
...
-    @expectedFailure
     @supported_platform
     @common_utils.parametrize("dtype", test_dtypes)
     def test_captured_buffers(self, dtype: torch.dtype):
-        head_offset = torch.rand(8, device="cuda", dtype=dtype)
+        head_offset = torch.rand(H, device="cuda", dtype=dtype)
 
         def score_mod(score, b, h, m, n):
-            return score + head_offset[h]
+            return score + index(head_offset, [h])
```
Comment (drisspg) [inline test/inductor/test_templated_attention.py:166]: I think we should make a follow up issue to figure out the indexing, maybe tangentially related: https://github.com/pytorch/pytorch/issues/124423

File: test/inductor/test_templated_attention.py [hunk 3]
```diff
@@ -147,23 +157,116 @@ def score_mod_2(score, b, h, m, n):
...
+                bidirectional_attention_on_video | causal_attention,
+                score,
+                -float("inf"),
+            )
+
+        self.run_test(bias_mod, dtype)
+
+    # @supported_platform
+    # @expectedFailure  # Triton bug?
+    # @common_utils.parametrize("dtype", test_dtypes)
```
Comment (drisspg) [inline test/inductor/test_templated_attention.py:244]: maybe uncomment?

File: torch/_higher_order_ops/templated_attention.py [hunk 4]
```diff
@@ -60,7 +60,7 @@ def math_attention(
     """
     assert len(other_buffers) == 0, "Other buffers are not yet supported."
 
-    scores = query @ key.transpose(-2, -1)
+    scores = (query @ key.transpose(-2, -1)).to(dtype=torch.float32)
```
Comment (drisspg) [inline torch/_higher_order_ops/templated_attention.py:63]: Idk about converting pre score_mod func

Comment (drisspg) [reply torch/_higher_order_ops/templated_attention.py:63]: Is that what the kernel's current logic is?

File: test/inductor/test_templated_attention.py [hunk 5]
```diff
@@ -39,6 +41,8 @@ def create_attention(score_mod):
     else [torch.float16, torch.float32]
 )
 
+test_dtypes_fast = [torch.float16]
```
Comment (drisspg) [inline test/inductor/test_templated_attention.py:44]: bfloat16, seems more important

File: test/inductor/test_templated_attention.py [hunk 6]
```diff
@@ -147,23 +157,116 @@ def score_mod_2(score, b, h, m, n):
...
+        num_bidirectional = torch.randint(0, S, (B,), device="cuda", dtype=torch.int32)
+
+        def bias_mod(score, b, h, q, kv):
+            causal_attention = q >= kv
+            cur_num_bidirectional = index(num_bidirectional, (b,))
+            bidirectional_attention_on_video = (q <= cur_num_bidirectional) & (
+                kv <= cur_num_bidirectional
+            )
+            return torch.where(
+                bidirectional_attention_on_video | causal_attention,
```
Comment (drisspg) [inline test/inductor/test_templated_attention.py:235]: could you also maybe try this with composition function?

**Review by drisspg — APPROVED** [review]
Very cool!

File: test/inductor/test_templated_attention.py (hunk 6 above)
Comment (Chillee) [reply test/inductor/test_templated_attention.py:235]: Composition here is actually a little bit tricky. Since you want the two masks to be "or"ed together and not "and"ed.

**Comment by Chillee** [comment]
@chillee has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56583900).
